v0/destinations/clickup/util.js

```diff
@@ -0,0 +1,383 @@
+ * https://api.clickup.com/api/v2/list/listId/field
+ * ref: https://clickup.com/api
+ * @param {*} listId
+ * @param {*} apiToken
+ * @returns
+ */
+const retrieveCustomFields = async (listId, apiToken) => {
+  const endpoint = getCustomFieldsEndPoint(listId);
+  const requestOptions = {
+    headers: {
+      "Content-Type": "application/json",
+      Authorization: apiToken
+    }
+  };
+  const customFieldsResponse = await httpGET(endpoint, requestOptions);
+  const processedCustomFieldsResponse = processAxiosResponse(
+    customFieldsResponse
+  );
+
+  if (processedCustomFieldsResponse.status !== 200) {
+    throw new CustomError(
+      `[ClickUp]:: Unable to retrieve available custom fields due to ${JSON.stringify(
+        processedCustomFieldsResponse.response
+      )}`,
+      processedCustomFieldsResponse.status
+    );
+  }
+
+  return processedCustomFieldsResponse.response?.fields;
+};
+
+/**
+ * Function to get the custom field value. As clickup is accepting uuid for dropdown/labels, and location in specific format
+ * @param {*} customField
+ * @param {*} value
+ * @returns
+ */
+const populateCustomFieldValue = (customField, value) => {
+  switch (customField?.type) {
+    case "drop_down":
+      return getDropDown(customField, value);
+    case "labels":
+      return getLabels(customField, value);
+    case "location":
+      return getLocation(value);
+    case "date":
+      return formatTimeStamp(value);
+    default:
+      return value;
+  }
+};
+
+/**
+ * Function to to build the custom field.
+ * returns the array of custom field with each object contains id,value
+ * @param {*} keyToCustomFieldName [{"from":"url","to":"Url"}]
+ * @param {*} properties {"url":"https://www.rudderstack.com/"}
+ * @param {*} listId
+ * @param {*} apiToken
+ * @returns [{"id":"b0f40a94-ea2a-4998-a514-8074d0eddcde","value":"https://www.rudderstack.com/"}]
+ */
+const customFieldsBuilder = async (
+  keyToCustomFieldName,
+  properties,
+  listId,
+  apiToken
+) => {
+  const responseArray = [];
+  if (properties && keyToCustomFieldName) {
+    // retrieve available clickup custom field for the given list
+    const availableCustomFields = await retrieveCustomFields(listId, apiToken);
+    // convert array to hashMap with key as field name and value as custom field object
+    const availableCustomFieldsMap = getHashFromArrayWithValueAsObject(
+      availableCustomFields,
+      "name",
+      false
+    );
+
+    // convert destination.Config.keyToCustomFieldName to hashMap
+    const configCustomFieldsMap = getHashFromArray(
+      keyToCustomFieldName,
+      "from",
+      "to",
+      false
+    );
+
+    Object.keys(configCustomFieldsMap).forEach(propertiesKey => {
+      let fieldValue = properties[propertiesKey];
+      if (fieldValue) {
+        const fieldName = configCustomFieldsMap[propertiesKey];
+        const customFieldDetailsObject = availableCustomFieldsMap[fieldName];
```

Comment:
`const customFieldDetailedInfo = extractUIMappedCustomFieldDetails (availableCustomFields, fieldName);`

Now call `getHashFromArrayWithValueAsObject` function inside this one and return `availableCustomFieldsMap[ fieldName ]`

Makes it more readable.

Reply:
rename `getHashFromArrayWithValueAsObject` as `createCustomFieldHashmap`

Reply:
> `const customFieldDetailedInfo = extractUIMappedCustomFieldDetails (availableCustomFields, fieldName);`
> 
> Now call `getHashFromArrayWithValueAsObject` function inside this one and return `availableCustomFieldsMap[ fieldName ]`
> 
> Makes it more readable.

Why do we need to call getHashFromArrayWithValueAsObject inside the extractUIMappedCustomFieldDetails??.
As, extractUIMappedCustomFieldDetails is being called inside the loop . We can call getHashFromArrayWithValueAsObject outside the loop that will the return the Hashmap from array. Then we can use the hashmap to get the required key/value.
Currently i have done the same on line no. 302.

Reply:
oh correct. I missed the loop scenario. My bad. Ok, no need to change the flow. Just update the variable names.

Reply:
Name changed and getHashFromArrayWithValueAsObject moved to global utils